components/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/token/AbstractJWTGenerator.java

```diff
@@ -0,0 +1,633 @@
+    }
+
+    /**
+     * Helper method to sign the JWT
+     *
+     * @param assertion
+     * @param endUserName
+     * @return signed assertion
+     * @throws org.wso2.carbon.appmgt.api.AppManagementException
+     */
+    private byte[] signJWT(String assertion, String endUserName)
+            throws AppManagementException {
+
+        try {
+            String tenantDomain = MultitenantUtils.getTenantDomain(endUserName);
+            int tenantId = getTenantId(endUserName);
+
+            Key privateKey = null;
+
+            if (!(privateKeys.containsKey(tenantId))) {
+                AppManagerUtil.loadTenantRegistry(tenantId);
+                //get tenant's key store manager
+                KeyStoreManager tenantKSM = KeyStoreManager.getInstance(tenantId);
+
+                if(!tenantDomain.equals(MultitenantConstants.SUPER_TENANT_DOMAIN_NAME)){
+                    //derive key store name
+                    String ksName = tenantDomain.trim().replace(".", "-");
+                    String jksName = ksName + ".jks";
+                    //obtain private key
+                    //TODO: maintain a hash map with tenants' private keys after first initialization
+                    privateKey = tenantKSM.getPrivateKey(jksName, tenantDomain);
+                }else{
+                    try{
+                        privateKey = tenantKSM.getDefaultPrivateKey();
+                    }catch (Exception e){
+                        log.error("Error while obtaining private key for super tenant",e);
+                    }
+                }
+                if (privateKey != null) {
+                    privateKeys.put(tenantId, privateKey);
+                }
+            } else {
+                privateKey = privateKeys.get(tenantId);
+            }
+
+            //initialize signature with private key and algorithm
+            Signature signature = Signature.getInstance(signatureAlgorithm);
+            signature.initSign((PrivateKey) privateKey);
+
+            //update signature with data to be signed
+            byte[] dataInBytes = assertion.getBytes();
+            signature.update(dataInBytes);
+
+            //sign the assertion and return the signature
+            byte[] signedInfo = signature.sign();
+            return signedInfo;
+
+        } catch (NoSuchAlgorithmException e) {
+            String error = "Signature algorithm not found.";
+            //do not log
+            throw new AppManagementException(error);
+        } catch (InvalidKeyException e) {
+            String error = "Invalid private key provided for the signature";
+            //do not log
+            throw new AppManagementException(error);
+        } catch (SignatureException e) {
+            String error = "Error in signature";
+            //do not log
+            throw new AppManagementException(error);
+        } catch (AppManagementException e) {
+            //do not log
+            throw new AppManagementException(e.getMessage());
```

Comment:
Add a message to the exception with some context info
Error in signature ${algorithm} for ${tenant}, ${what}